File: reader-app/src/reader/narration/paragraph-builder.ts
import { LINE_Y_TOLERANCE, PARAGRAPH_GAP_FACTOR } from "../core/config.ts";
import type { BoundingBox, DocumentBlock } from "../core/types.ts";

const PAGE_NUMBER_RE =
  /^(?:page\s+)?(?:[ivxlcdm]+|\d+)(?:\s*\/\s*\d+)?$|^[ivxlcdm]+$|^\d+\s*$/i;

export function isPageNumber(text: string): boolean {
  const trimmed = text.trim();
  if (trimmed.length === 0 || trimmed.length > 18) return false;
  return PAGE_NUMBER_RE.test(trimmed);
}

export interface TextLine {
  text: string;
  blocks: DocumentBlock[];
  bounds: BoundingBox;
  fontSize: number;
}

export interface ParagraphGroup {
  text: string;
  originalParts: string[];
  sourceLines: TextLine[];
  blocks: DocumentBlock[];
  bounds: BoundingBox[];
  fontSize: number;
  indent: number;
}

function unionBounds(boxes: BoundingBox[]): BoundingBox {
  let minX = 1;
  let minY = 1;
  let maxX = 0;
  let maxY = 0;
  for (const box of boxes) {
    minX = Math.min(minX, box.x);
    minY = Math.min(minY, box.y);
    maxX = Math.max(maxX, box.x + box.width);
    maxY = Math.max(maxY, box.y + box.height);
  }
  return {
    x: minX,
    y: minY,
    width: Math.max(0, maxX - minX),
    height: Math.max(0, maxY - minY),
  };
}

// Decorative markers - section bullets, drop-cap rules, dingbats - are set in a
// larger face than the words they sit beside. Letting them into the line box
// pushes the reading highlight off the text, so the box follows the blocks that
// actually carry words whenever the line has any.
function wordBearing(blocks: DocumentBlock[]): DocumentBlock[] {
  const words = blocks.filter((block) => /[\p{L}\p{N}]/u.test(block.text));
  return words.length > 0 ? words : blocks;
}

function lineBounds(blocks: DocumentBlock[]): BoundingBox {
  return unionBounds(wordBearing(blocks).map((block) => block.bounds));
}

// A running page-number stamp (roman or arabic) sometimes shares a line with
// a chapter heading instead of sitting alone in the header/footer band. Most
// books set it well below the heading's font size, unlike a real word the
// heading would use at full size, so a small isolated page-number token is
// dropped from the spoken line. Some books set the folio at the heading's own
// size instead, so a stamp is also dropped when it sits apart from its
// nearest neighbor by a much wider gap than the line's own word spacing -
// true of a folio parked in the margin, not of a word inside the heading.
function dropStrayPageNumbers(blocks: DocumentBlock[]): DocumentBlock[] {
  if (blocks.length < 2) return blocks;
  const dominant = Math.max(...blocks.map((block) => block.fontSize ?? 0));
  if (dominant <= 0) return blocks;
  const filtered = blocks.filter((block, index) => {
    if (!isPageNumber(block.text)) return true;
    const size = block.fontSize ?? dominant;
    const isSmallStamp = size < dominant * 0.7;
    const isIsolatedStamp = isFarFromNeighbors(blocks, index);
    return !(isSmallStamp || isIsolatedStamp);
  });
  return filtered.length > 0 ? filtered : blocks;
}

function gapBetween(a: DocumentBlock, b: DocumentBlock): number {
  return Math.max(0, b.bounds.x - (a.bounds.x + a.bounds.width));
}

// A folio stamp set apart from the rest of the line sits behind a gap several
// character-widths wider than ordinary word spacing, regardless of the page's
// absolute scale. A block's own font size is a stand-in for its character
// width, since normalized page coordinates carry no font metrics directly.
const STRAY_GAP_EMS = 2.5;

function isFarFromNeighbors(blocks: DocumentBlock[], index: number): boolean {
  const block = blocks[index]!;
  const em = block.fontSize ?? block.bounds.height;
  const threshold = em * STRAY_GAP_EMS;
  const before = index > 0 ? gapBetween(blocks[index - 1]!, block) : Infinity;
  const after = index < blocks.length - 1 ? gapBetween(block, blocks[index + 1]!) : Infinity;
  return Math.min(before, after) >= threshold;
}

// Letter-tracked display type (title pages, covers) reaches pdf.js as text
// where every glyph is its own space-separated token, e.g. "P e n g u i n".
// Read as-is, a TTS engine spells the word out letter by letter, so runs of
// single-character tokens are collapsed back into one word before joining.
function collapseLetterTracking(text: string): string {
  return text.replace(
    /(?:^|(?<=\s))\p{L}(?:\s\p{L}){2,}(?=\s|$)/gu,
    (run) => run.replace(/\s+/g, ""),
  );
}

function joinLineText(rawBlocks: DocumentBlock[], preserveFolio: boolean): string {
  const blocks = preserveFolio ? rawBlocks : dropStrayPageNumbers(rawBlocks);
  const joined = blocks
    .map((block) => block.text.replace(/\s+/g, " ").trim())
    .filter(Boolean)
    .join(" ")
    .replace(/\s+/g, " ")
    .trim();
  return collapseLetterTracking(joined);
}

function isDropCapContinuation(line: TextLine, block: DocumentBlock): boolean {
  if (line.blocks.length !== 1 || !/^\p{Lu}$/u.test(line.text.trim())) return false;
  const cap = line.blocks[0];
  if (!cap) return false;
  const bodySize = block.fontSize ?? block.bounds.height;
  const capSize = cap.fontSize ?? cap.bounds.height;
  const capBottom = cap.bounds.y + cap.bounds.height;
  const horizontalGap = block.bounds.x - (cap.bounds.x + cap.bounds.width);
  return capSize >= bodySize * 1.6 &&
    horizontalGap >= 0 && horizontalGap <= bodySize * 1.5 &&
    cap.bounds.y <= block.bounds.y + block.bounds.height / 2 &&
    capBottom >= block.bounds.y;
}

export function groupLines(
  blocks: DocumentBlock[],
  options: { preserveFolio?: boolean } = {},
): TextLine[] {
  const preserveFolio = options.preserveFolio ?? false;
  const lines: TextLine[] = [];
  for (const block of blocks) {
    const last = lines.at(-1);
    if (!last) {
      lines.push({
        text: joinLineText([block], preserveFolio),
        blocks: [block],
        bounds: { ...block.bounds },
        fontSize: block.fontSize ?? block.bounds.height,
      });
      continue;
    }

    const lastSize = last.fontSize || 0.02;
    const yClose =
      Math.abs(
        block.bounds.y + block.bounds.height / 2 - (last.bounds.y + last.bounds.height / 2),
      ) <=
      lastSize * LINE_Y_TOLERANCE;
    const sizeClose = Math.abs((block.fontSize ?? lastSize) - lastSize) <= lastSize * 0.45;

    if (yClose && sizeClose) {
      last.blocks.push(block);
      last.bounds = lineBounds(last.blocks);
      last.text = joinLineText(last.blocks, preserveFolio);
      last.fontSize = Math.max(last.fontSize, block.fontSize ?? last.fontSize);
    } else if (isDropCapContinuation(last, block)) {
      last.blocks.push(block);
      last.bounds = unionBounds(last.blocks.map((item) => item.bounds));
      last.text = joinLineText(last.blocks, preserveFolio);
      last.fontSize = block.fontSize ?? block.bounds.height;
    } else {
      lines.push({
        text: joinLineText([block], preserveFolio),
        blocks: [block],
        bounds: { ...block.bounds },
        fontSize: block.fontSize ?? block.bounds.height,
      });
    }
  }
  return lines.filter(
    (line) => line.text.trim().length > 0 && (preserveFolio || !isFolioLine(line)),
  );
}

// A line made up of nothing but a page-number stamp (optionally beside a
// decorative rule or bullet) is a running folio, not prose - regardless of
// which document it came from. Left in, it merges into whichever paragraph
// sits on the next line (often a chapter heading) and gets read aloud with it.
function isFolioLine(line: TextLine): boolean {
  const prose = wordBearing(line.blocks).filter((block) =>
    /\p{L}{2,}/u.test(block.text),
  );
  if (prose.length !== 0) return false;
  const words = line.blocks.filter((block) => /[\p{L}\p{N}]/u.test(block.text));
  return words.length > 0 && words.every((block) => isPageNumber(block.text));
}

const KEEP_HYPHEN_NEXT = new Set(["of", "the", "and", "to", "in", "for", "with", "or", "a", "an"]);

export function dehyphenate(previous: string, next: string): string {
  const match = previous.match(/^(.*?)([A-Za-z]{2,})-$/);
  if (!match) {
    if (previous.endsWith("-")) return `${previous}${next}`;
    return `${previous} ${next}`;
  }
  const stem = match[1] ?? "";
  const prefix = match[2] ?? "";
  const nextWord = (next.match(/^[A-Za-z']+/)?.[0] ?? "").toLowerCase();
  if (/^[a-z]/.test(next) && prefix.length >= 4 && !KEEP_HYPHEN_NEXT.has(nextWord)) {
    return `${stem}${prefix}${next}`;
  }
  return `${stem}${prefix}-${next}`;
}

function looksLikeList(text: string): boolean {
  return /^(?:[-•●▪]|[0-9]{1,2}[.)]|[A-Za-z][.)])\s+\S/.test(text.trim());
}

function looksLikeBreak(
  line: { bounds: BoundingBox; fontSize: number; text: string },
  next: { bounds: BoundingBox; fontSize: number; text: string },
): boolean {
  const gap = next.bounds.y - (line.bounds.y + line.bounds.height);
  const size = line.fontSize || 0.02;
  if (gap > size * PARAGRAPH_GAP_FACTOR) return true;
  if (Math.abs(next.bounds.x - line.bounds.x) > 0.08) return true;
  // A ~15%+ font-size change reliably marks a new heading or label starting
  // (real headings commonly run only ~15-20% larger than body text - a 35%
  // threshold missed those and let a heading merge into the paragraph after it).
  if (Math.abs(next.fontSize - line.fontSize) > line.fontSize * 0.15) return true;
  if (looksLikeList(next.text)) return true;
  if (/[:]$/.test(line.text) && looksLikeList(next.text)) return true;
  if (/^\[(?:\d+(?:[,;–-]\s*\d+)*)\]$/.test(line.text.trim())) return true;
  const ratio = line.bounds.width === 0 ? 1 : line.bounds.width;
  if (line.bounds.width < 0.35 && /[.!?]$/.test(line.text) && ratio < 0.45) {
    return true;
  }
  return false;
}

export function groupParagraphs(lines: TextLine[]): ParagraphGroup[] {
  const groups: ParagraphGroup[] = [];
  for (const line of lines) {
    const last = groups.at(-1);
    if (!last) {
      groups.push({
        text: line.text.trim(),
        originalParts: [line.text.trim()],
        sourceLines: [line],
        blocks: [...line.blocks],
        bounds: [line.bounds],
        fontSize: line.fontSize,
        indent: line.bounds.x,
      });
      continue;
    }

    const lastLine = {
      bounds: last.bounds.at(-1) ?? line.bounds,
      fontSize: last.fontSize,
      text: last.originalParts.at(-1) ?? last.text,
    };

    if (looksLikeBreak(lastLine, line) || looksLikeList(line.text)) {
      groups.push({
        text: line.text.trim(),
        originalParts: [line.text.trim()],
        sourceLines: [line],
        blocks: [...line.blocks],
        bounds: [line.bounds],
        fontSize: line.fontSize,
        indent: line.bounds.x,
      });
      continue;
    }

    last.text = dehyphenate(last.text, line.text.trim());
    last.originalParts.push(line.text.trim());
    last.sourceLines.push(line);
    last.blocks.push(...line.blocks);
    last.bounds.push(line.bounds);
    last.fontSize = Math.max(last.fontSize, line.fontSize);
  }
  return groups;
}

File: android-app/core/src/main/kotlin/com/evangeline/reader/narration/ParagraphBuilder.kt
package com.evangeline.reader.narration

import com.evangeline.reader.model.BoundingBox
import com.evangeline.reader.model.DocumentBlock
import com.evangeline.reader.model.LINE_Y_TOLERANCE
import com.evangeline.reader.model.PARAGRAPH_GAP_FACTOR
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/** One visual line of the page: the runs of text sharing a baseline. */
data class TextLine(
    val text: String,
    val blocks: List<DocumentBlock>,
    val bounds: BoundingBox,
    val fontSize: Double,
)

private const val FALLBACK_FONT_SIZE = 0.02
private const val SIZE_TOLERANCE = 0.45
private val WHITESPACE = Regex("\\s+")
private val WORD_CHARACTER = Regex("[\\p{L}\\p{N}]")

private fun unionBounds(boxes: List<BoundingBox>): BoundingBox {
    var minX = 1.0
    var minY = 1.0
    var maxX = 0.0
    var maxY = 0.0
    for (box in boxes) {
        minX = min(minX, box.x)
        minY = min(minY, box.y)
        maxX = max(maxX, box.x + box.width)
        maxY = max(maxY, box.y + box.height)
    }
    return BoundingBox(minX, minY, max(0.0, maxX - minX), max(0.0, maxY - minY))
}

// Decorative markers - section bullets, drop-cap rules, dingbats - are set in a
// larger face than the words beside them. Letting them into the line box pushes
// the reading highlight off the text, so the box follows the blocks that
// actually carry words whenever the line has any.
private fun wordBearing(blocks: List<DocumentBlock>): List<DocumentBlock> {
    val words = blocks.filter { WORD_CHARACTER.containsMatchIn(it.text) }
    return words.ifEmpty { blocks }
}

private fun lineBounds(blocks: List<DocumentBlock>): BoundingBox =
    unionBounds(wordBearing(blocks).map { it.bounds })

private fun joinLineText(blocks: List<DocumentBlock>): String =
    blocks
        .map { it.text.replace(WHITESPACE, " ").trim() }
        .filter { it.isNotEmpty() }
        .joinToString(" ")
        .replace(WHITESPACE, " ")
        .trim()

private fun DocumentBlock.sizeOrHeight(): Double = fontSize ?: bounds.height

private fun startLine(block: DocumentBlock) = TextLine(
    text = block.text,
    blocks = listOf(block),
    bounds = block.bounds,
    fontSize = block.sizeOrHeight(),
)

private fun BoundingBox.verticalCentre(): Double = y + height / 2

private fun joinsLine(line: TextLine, block: DocumentBlock): Boolean {
    val lastSize = if (line.fontSize == 0.0) FALLBACK_FONT_SIZE else line.fontSize
    val centreGap = abs(block.bounds.verticalCentre() - line.bounds.verticalCentre())
    val sizeGap = abs((block.fontSize ?: lastSize) - lastSize)
    return centreGap <= lastSize * LINE_Y_TOLERANCE && sizeGap <= lastSize * SIZE_TOLERANCE
}

private fun extend(line: TextLine, block: DocumentBlock): TextLine {
    val blocks = line.blocks + block
    return TextLine(
        text = joinLineText(blocks),
        blocks = blocks,
        bounds = lineBounds(blocks),
        fontSize = max(line.fontSize, block.fontSize ?: line.fontSize),
    )
}

/**
 * Groups positioned text runs into visual lines.
 *
 * A run joins the current line when its vertical centre and font size are both
 * close enough; otherwise it starts a new line. Empty lines are dropped so they
 * never become silent passages.
 */
fun groupLines(blocks: List<DocumentBlock>): List<TextLine> {
    val lines = mutableListOf<TextLine>()
    for (block in blocks) {
        val last = lines.lastOrNull()
        if (last != null && joinsLine(last, block)) {
            lines[lines.lastIndex] = extend(last, block)
        } else {
            lines.add(startLine(block))
        }
    }
    return lines.filter { it.text.isNotBlank() }
}

// Short words that follow a hyphen are usually a real compound ("state-of-the-art"),
// not a word broken across lines, so the hyphen is kept.
private val KEEP_HYPHEN_NEXT = setOf(
    "of", "the", "and", "to", "in", "for", "with", "or", "a", "an",
)

private val HYPHEN_BREAK = Regex("^(.*?)([A-Za-z]{2,})-$")
private val LEADING_WORD = Regex("^[A-Za-z']+")
private const val MIN_JOINABLE_PREFIX = 4

/**
 * Joins a line ending in a hyphen to the line after it.
 *
 * The hyphen is dropped only when the break looks like one word split across
 * lines: a prefix of at least four letters, continued by a lowercase word that
 * is not a common short word. Anything else keeps its hyphen, because removing
 * it from a real compound would change what the listener hears.
 */
fun dehyphenate(previous: String, next: String): String {
    val match = HYPHEN_BREAK.find(previous)
        ?: return if (previous.endsWith("-")) "$previous$next" else "$previous $next"

    val stem = match.groupValues[1]
    val prefix = match.groupValues[2]
    val nextWord = LEADING_WORD.find(next)?.value.orEmpty().lowercase()
    val continuesWord = next.firstOrNull()?.isLowerCase() == true &&
        prefix.length >= MIN_JOINABLE_PREFIX &&
        nextWord !in KEEP_HYPHEN_NEXT

    return if (continuesWord) "$stem$prefix$next" else "$stem$prefix-$next"
}

/**
 * A run of lines that belong together as one block of prose.
 *
 * [text] is the joined, de-hyphenated reading text; [originalParts] keeps each
 * line as it appeared, because the highlight maps to lines on the page while
 * narration speaks the joined form.
 */
data class ParagraphGroup(
    val text: String,
    val originalParts: List<String>,
    val sourceLines: List<TextLine>,
    val blocks: List<DocumentBlock>,
    val bounds: List<BoundingBox>,
    val fontSize: Double,
    val indent: Double,
)

private val LIST_START = Regex("^(?:[-•●▪]|[0-9]{1,2}[.)]|[A-Za-z][.)])\\s+\\S")
private val SENTENCE_END = Regex("[.!?]$")
private const val INDENT_BREAK = 0.08
private const val SIZE_BREAK = 0.35
private const val SHORT_LINE_WIDTH = 0.35
private const val SHORT_LINE_RATIO = 0.45

private fun looksLikeList(text: String): Boolean = LIST_START.containsMatchIn(text.trim())

// A line ending a paragraph is usually short and finishes a sentence; a line
// that fills its column is usually mid-paragraph.
private fun endsBlock(line: TextLine): Boolean =
    line.bounds.width < SHORT_LINE_WIDTH &&
        SENTENCE_END.containsMatchIn(line.text) &&
        (if (line.bounds.width == 0.0) 1.0 else line.bounds.width) < SHORT_LINE_RATIO

private fun looksLikeBreak(line: TextLine, next: TextLine): Boolean {
    val size = if (line.fontSize == 0.0) FALLBACK_FONT_SIZE else line.fontSize
    val gap = next.bounds.y - (line.bounds.y + line.bounds.height)
    return gap > size * PARAGRAPH_GAP_FACTOR ||
        abs(next.bounds.x - line.bounds.x) > INDENT_BREAK ||
        abs(next.fontSize - line.fontSize) > line.fontSize * SIZE_BREAK ||
        looksLikeList(next.text) ||
        endsBlock(line)
}

private fun startGroup(line: TextLine) = ParagraphGroup(
    text = line.text.trim(),
    originalParts = listOf(line.text.trim()),
    sourceLines = listOf(line),
    blocks = line.blocks,
    bounds = listOf(line.bounds),
    fontSize = line.fontSize,
    indent = line.bounds.x,
)

private fun extend(group: ParagraphGroup, line: TextLine) = group.copy(
    text = dehyphenate(group.text, line.text.trim()),
    originalParts = group.originalParts + line.text.trim(),
    sourceLines = group.sourceLines + line,
    blocks = group.blocks + line.blocks,
    bounds = group.bounds + line.bounds,
    fontSize = max(group.fontSize, line.fontSize),
)

/** Groups visual lines into paragraphs, joining words broken across lines. */
fun groupParagraphs(lines: List<TextLine>): List<ParagraphGroup> {
    val groups = mutableListOf<ParagraphGroup>()
    for (line in lines) {
        val last = groups.lastOrNull()
        if (last == null || looksLikeBreak(last.sourceLines.last(), line) || looksLikeList(line.text)) {
            groups.add(startGroup(line))
        } else {
            groups[groups.lastIndex] = extend(last, line)
        }
    }
    return groups
}
